src/Columns/ColumnDynamic.cpp

```diff
@@ -979,6 +979,92 @@ ColumnPtr ColumnDynamic::compress() const
         });
 }
 
+void ColumnDynamic::updateCheckpoint(ColumnCheckpoint & checkpoint) const
+{
+    auto & nested = assert_cast<ColumnCheckpointWithMultipleNested &>(checkpoint).nested;
+    const auto & variants = variant_column_ptr->getVariants();
+
+    size_t old_size = nested.size();
+    chassert(old_size <= variants.size());
+
+    for (size_t i = 0; i < old_size; ++i)
+    {
+        variants[i]->updateCheckpoint(*nested[i]);
+    }
+
+    /// If column has new variants since last checkpoint create checkpoints for them.
+    if (old_size < variants.size())
+    {
+        nested.resize(variants.size());
+        for (size_t i = old_size; i < variants.size(); ++i)
+            nested[i] = variants[i]->getCheckpoint();
+    }
+
+    checkpoint.size = size();
+}
+
+
+DataTypePtr ColumnDynamic::popBackVariants(const VariantInfo & info, const std::vector<ColumnVariant::Discriminator> & local_to_global_discriminators, size_t n)
+{
+    const auto & type_variant = assert_cast<const DataTypeVariant &>(*info.variant_type);
+
+    std::unordered_map<ColumnVariant::Discriminator, String> discriminator_to_name;
+    std::unordered_map<String, DataTypePtr> name_to_data_type;
+
+    for (const auto & [name, discriminator] : info.variant_name_to_discriminator)
+        discriminator_to_name.emplace(discriminator, name);
+
+    for (const auto & type : type_variant.getVariants())
+        name_to_data_type.emplace(type->getName(), type);
```

Comment:
Why not to create single map `discriminator -> type` by iterating over `type_variant.getVariants()`?
`type_variant.getVariants()` returns variants in global order